systemd/watch.go

```diff
@@ -5,65 +5,77 @@ import (
 	"sync"
 
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog/v2"
 )
 
-// The Updater interface we will send notifications when configmaps and secrets are updated.
-type Updater interface {
-	updateConfigMap(context.Context, *corev1.Pod, *corev1.ConfigMap) error
-	updateSecret(context.Context, *corev1.Pod, *corev1.Secret) error
+// The updater notifies pods when related configmaps and secrets change.
+type updater interface {
+	updateConfigMap(ctx context.Context, pod *corev1.Pod, configMap *corev1.ConfigMap) error
+	updateSecret(ctx context.Context, pod *corev1.Pod, secret *corev1.Secret) error
 }
 
-// Watcher checks the API server for configMap and secret updates and notifies the provider.
-type Watcher struct {
+// watcher checks the API server for configMap and secret updates and notifies the provider.
+type watcher struct {
 	mu sync.RWMutex
-	// maps are indexed by "namespace.name"
-	configs map[string][]*corev1.Pod
-	secrets map[string][]*corev1.Pod
+	// must keep a deep-copy of Pod since its volumes are gone when deletion is triggered.
+	configs map[types.NamespacedName][]*corev1.Pod
+	// cmKeysByPod enables reverse lookup ConfigMaps keys per Pod.
+	cmKeysByPod map[types.NamespacedName][]types.NamespacedName
+
+	// must keep a deep-copy of Pod since its volumes are gone when deletion is triggered.
+	secrets map[types.NamespacedName][]*corev1.Pod
+	// secretKeysByPod enables reverse lookup Secrets keys per Pod.
+	secretKeysByPod map[types.NamespacedName][]types.NamespacedName
 }
 
-func newWatcher() *Watcher {
-	return &Watcher{
-		configs: make(map[string][]*corev1.Pod),
-		secrets: make(map[string][]*corev1.Pod),
+func newWatcher() *watcher {
+	return &watcher{
+		configs:         make(map[types.NamespacedName][]*corev1.Pod),
+		cmKeysByPod:     make(map[types.NamespacedName][]types.NamespacedName),
+		secrets:         make(map[types.NamespacedName][]*corev1.Pod),
+		secretKeysByPod: make(map[types.NamespacedName][]types.NamespacedName),
 	}
 }
 
-func (w *Watcher) handlerFuncs(p Updater) cache.ResourceEventHandlerFuncs {
+func (w *watcher) handlerFuncs(ctx context.Context, p updater) cache.ResourceEventHandlerFuncs {
 	return cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
-			w.handleEvent(obj, p)
+			w.handleEvent(ctx, obj, p)
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
-			w.handleEvent(newObj, p)
+			w.handleEvent(ctx, newObj, p)
 		},
 		DeleteFunc: nil, // Ignore for now. Pod deletion takes care of cleaning up.
 	}
 }
-func (w *Watcher) handleEvent(obj interface{}, upd Updater) {
+func (w *watcher) handleEvent(ctx context.Context, obj interface{}, upd updater) {
 	switch v := obj.(type) {
-	case *corev1.Secret:
+	case *corev1.ConfigMap:
```

Comment:
why the reversal of these cases?

Reply:
To normalize code. In every other func that handles both ConfigMaps and Secrets, ConfigMaps are handled first.